org/postgresql/core/NativeQuery.java

```diff
@@ -0,0 +1,77 @@
+/*-------------------------------------------------------------------------
+*
+* Copyright (c) 2003-2014, PostgreSQL Global Development Group
+* Copyright (c) 2004, Open Cloud Limited.
+*
+*
+*-------------------------------------------------------------------------
+*/
+package org.postgresql.core;
+
+/**
+ * Represents a query that is ready for execution by backend.
+ * The main difference from JDBC is ? are replaced with $1, $2, etc.
+ */
+public class NativeQuery {
+    private final static String[] BIND_NAMES = new String[128];
+    private final static int[] NO_BINDS = new int[0];
+
+    public final String nativeSql;
+    public final int[] bindPositions;
+
+    static
+    {
+        for (int i = 1; i < BIND_NAMES.length; i++)
+        {
+            BIND_NAMES[i] = "$" + i;
+        }
+    }
+
+    public NativeQuery(String nativeSql)
+    {
+        this(nativeSql, NO_BINDS);
+    }
+
+    public NativeQuery(String nativeSql, int[] bindPositions)
+    {
+        this.nativeSql = nativeSql;
+        this.bindPositions = bindPositions == null || bindPositions.length == 0 ? NO_BINDS : bindPositions;
+    }
+
+    /**
+     * Stringize this query to a human-readable form, substituting
+     * particular parameter values for parameter placeholders.
+     *
+     * @param parameters a ParameterList returned by this Query's
+     *  {@link Query#createParameterList} method, or <code>null</code> to
+     *  leave the parameter placeholders unsubstituted.
+     * @return a human-readable representation of this query
+     */
+    public String toString(ParameterList parameters)
+    {
+        if (bindPositions.length == 0)
+            return nativeSql;
+
+        StringBuilder sbuf = new StringBuilder(nativeSql.length());
```

Comment:
Something like `nativeSql.length() + 2 * bindPositions.length` might save a `StringBuilder` resize in the common case?